aip/general/0193.md

```diff
@@ -75,224 +84,354 @@ following:
 
-When including `LocalizedMessage`, both fields, `locale` and `message`,
-**must** be populated.  If the service is to be localized, the value of
-`locale` **must** change dynamically. See,
-"[Localization](#localization)". Otherwise, `locale` **must** always
-present the service's default locale, e.g. "en-US".
+**Note:** `ErrorInfo` represents a special case. There **must** be
+exactly one `ErrorInfo`. It is required.
 
-When adding an error message via `LocalizedMessage`, `ErrorInfo`
-**must** be introduced either before or at the same time. If there are
-dynamic segments found in the text, the values of these variables
-**must** be included in `ErrorInfo.metadata`.  See, "[Dynamic
-variables](#dynamic-variables)".
+[ErrorInfo Payload][ErrorInfo]
 
-**Warning:** If `LocalizedMessage` is released without `ErrorInfo`, the
-same [concerns](#status-message-warning) regarding client misuse of
-textual error messages apply.
+Details of this object are summarized in the following fields, field
+descriptions, and examples:
 
-#### Dynamic variables
++ `reason string`
+  + A short snake_case description of why the error occurred. Error reasons are
+    unique within a particular domain of errors. The error reason **must** do the following:
+  + Be at most 63 characters and match a regular expression of
+    \`\[A-Z\]\[A-Z0-9\_\]+\[A-Z0-9\]\`, which represents UPPER_SNAKE_CASE.
+  + Be unique and consumable by machine actors for automation.
+  + Be meaningful enough for a human reader to understand what the reason 
+    refers  to.
+  + Map to one and only one cause, That is, avoid making machine actors reliant
+    on ErrorInfo.
+  + *Example*: CPU_AVAILABILITY
+  + Distill your error message into its simplest form. For example, if your
+    error  message is: `The Book, "The Great Gatsby", is unavailable at the
+    Library, "Garfield East". It is expected to be available again on
+    2199-05-13.` The `reason string` could be one of the following text examples in UPPER_SNAKE_CASE: UNAVAILABLE, NO_STOCK, CHECKED_OUT, AVAILABILITY_ERROR
+  + In contrast, using any of the following reasons breaches the 
+    intended formatting and is not recommended: noBooks, THE_BOOK_YOU_WANT_IS_NOT_AVAILABLE, ERROR, librariesAreGreat
++ `domain string`
+  + The logical grouping to which the `reason` belongs. The error domain is
+    typically the registered service name of the tool or product that generated the error. The domain must be a globally unique value.
+  + *Example*: "`pubsub.googleapis.com`"
++ `metadata map`
+  + Additional structured details about this error, which **should** provide
+    important context for customers to identify resolution steps.  Keys
+    **should** match /\[a-zA-Z0-9-\_\]/ and be limited to 64 characters in
```

Comment:
the regex here seems wrong, "us-", "us_", "us-_" are all valid keys according to it. I can help craft a better one, what are the constraints here?

Reply:
@sgiovanni1  It looks like the constraints for metadata are using upper or lower case letters, or numbers, Stanis, can you please verify this?

Reply:
the regex here must match the one in the proto. please use `/[a-z][a-zA-Z0-9-_]+/`, we're in the process of updating the public proto documentation to match this: https://github.com/googleapis/googleapis/blob/fe20507f2a5a17da501cd3bcd4dfbc7675fb5a3f/google/rpc/error_details.proto#L69

Reply:
Updated.